### ISSUE TITLE
Add a check to prevent executing espanso with root permissions on macOS 

### ISSUE THREAD

**Issue by mmynsted** [federico-terzi/espanso#648]
I installed espanso espanso `0.7.3` using homebrew on macOS `10.15.7`. I followed the install instructions and all seemed well until I tried to start espanso. When I start espanso I get a permission error for creating a lock file. What should I do?

```bash
11:00 $ espanso start
thread 'main' panicked at 'Cannot create reference to lock file.: Os { code: 13, kind: PermissionDenied, message: "Permission denied" }', src/main.rs:1426:10
stack backtrace:
   0:        0x10ce14194 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::hcfc48256a5ab8835
   1:        0x10ce397c0 - core::fmt::write::haf3903118f694c48
   2:        0x10ce0c926 - std::io::Write::write_fmt::h7385463ac87804ed
   3:        0x10ce165af - std::panicking::default_hook::{{closure}}::h91bd4c58cf71392b
   4:        0x10ce1627d - std::panicking::default_hook::h7bd29c87df967048
   5:        0x10ce16b7b - std::panicking::rust_panic_with_hook::hae2b05f08a320721
   6:        0x10ce166fb - std::panicking::begin_panic_handler::{{closure}}::h72d68d3a77e0b718
   7:        0x10ce14608 - std::sys_common::backtrace::__rust_end_short_backtrace::h7c5e286792f94edb
   8:        0x10ce166ba - _rust_begin_unwind
   9:        0x10ce55f9f - core::panicking::panic_fmt::h1b194bb80d76fb10
  10:        0x10ce55ea5 - core::result::unwrap_failed::h56c9cad3598d6b68
  11:        0x10ca0af7c - espanso::acquire_lock::hc51210a49c35db45
  12:        0x10ca0907c - espanso::start_main::hfe4bc120e9a9cc2b
  13:        0x10ca017aa - espanso::main::hbcf7755afdb51899
  14:        0x10c97fe0a - std::sys_common::backtrace::__rust_begin_short_backtrace::he96e5589e895f929
  15:        0x10c9a91dc - std::rt::lang_start::{{closure}}::h83941110f3b2edad
  16:        0x10ce16fd4 - std::rt::lang_start_internal::hd38bb63f9540b327
  17:        0x10ca0b6d9 - _main
```

**Comment by federico-terzi** [comment]
Hey @mmynsted,

I'm sorry you experienced this problem! I still haven't had the chance to properly test espanso on Big Sur, so it might be related...
Could you please try granting the Terminal Full Disk Access using these steps?

1. Navigate to Apple Menu › System Preferences › Security & Privacy › Privacy › Full Disk Access
2. Enable the terminal there

Let me know if that helps!

Cheers :)

**Comment by mmynsted** [comment]
When I look at the code it appears that a lock file is being created in the data directory but the it fails from a lack of permission. 

The data directory on my system appears to be `/Users/mmynsted/Library/Application\ Support/espanso`. I checked and somehow this directory is owned by `root`. 

I suspect there is an issue with the homebrew install. 

The following change permitted `espanso` to start. 

```bash
sudo chown -R mmynsted /Users/mmynsted/Library/Application\ Support/espanso\
```

**Comment by federico-terzi** [comment]
@mmynsted Thank you for the investigation! Glad you got it working

> The data directory on my system is /Users/mmynsted/Library/Application\ Support/espanso. I checked and somehow this directory is owned by root.

This is interesting... Perhaps the first time you called "espanso register" you used sudo with `sudo espanso register`?

**Comment by mmynsted** [comment]
It is possible I registered it using root by accident. Looks like there are other issues. 

```bash
$ espanso start
Daemon started correctly!
```

But 

```bash
$ espanso status
espanso is not running
```

I tried uninstalling and re-installing but that has not fixed things.

**Comment by federico-terzi** [comment]
@mmynsted I see, thank you. Could you please post the output of `espanso log` here?

**Comment by mmynsted** [comment]
```bash
$ espanso log
No log file found.
```

**Comment by federico-terzi** [comment]
@mmynsted Thank you, this is getting interesting... There might be other permission problems. Could you please run the following:

```
espanso -vv daemon
```

And then post the output here? (you might need to terminate it with CTRL+C after a few seconds)

**Comment by mmynsted** [comment]
```bash
$ espanso -vv daemon
16:37:47 [ INFO] espanso version 0.7.3
16:37:47 [ INFO] using config path: /Users/mmynsted/Library/Preferences/espanso
16:37:47 [ INFO] using package path: /Users/mmynsted/Library/Application Support/espanso/packages
16:37:47 [ INFO] spawning worker process...
16:37:47 [ WARN] Unable to delete Unix socket: No such file or directory (os error 2)
16:37:47 [ INFO] Binded to IPC unix socket: /Users/mmynsted/Library/Application Support/espanso/espanso.sock
16:37:47 [ WARN] Unable to delete Unix socket: No such file or directory (os error 2)
16:37:47 [ INFO] watching for changes in path: /Users/mmynsted/Library/Preferences/espanso
2021-04-11 11:37:48.093 EspansoNotifyHelper[58138:15319753] TISFileInterrogator updateSystemInputSources false but old data invalid: currentCacheHeaderPtr nonNULL? 0, ->cacheFormatVersion 0, ->magicCookie 00000000, inputSourceTableCountSys 0
Keyboard Layouts: duplicate keyboard layout identifier -17410.
Keyboard Layouts: keyboard layout identifier -17410 has been replaced with -28673.
Keyboard Layouts: duplicate keyboard layout identifier -30769.
Keyboard Layouts: keyboard layout identifier -30769 has been replaced with -28674.
Keyboard Layouts: duplicate keyboard layout identifier -14934.
Keyboard Layouts: keyboard layout identifier -14934 has been replaced with -28675.
```

**Comment by mmynsted** [comment]
```bash
$ espanso log
16:37:47 [ INFO] espanso version 0.7.3
16:37:47 [ INFO] using config path: /Users/mmynsted/Library/Preferences/espanso
16:37:47 [ INFO] using package path: /Users/mmynsted/Library/Application Support/espanso/packages
16:37:47 [ INFO] spawning worker process...
16:37:47 [ WARN] Unable to delete Unix socket: No such file or directory (os error 2)
16:37:47 [ INFO] Binded to IPC unix socket: /Users/mmynsted/Library/Application Support/espanso/espanso.sock
16:37:47 [ INFO] initializing worker process...
16:37:47 [ INFO] Initializing EspansoNotifyHelper in /Users/mmynsted/Library/Application Support/espanso
16:37:47 [ WARN] Unable to delete Unix socket: No such file or directory (os error 2)
16:37:47 [ INFO] Binded to IPC unix socket: /Users/mmynsted/Library/Application Support/espanso/worker.sock
16:37:47 [ INFO] Using modulo at "/usr/local/bin/modulo"
16:37:47 [ INFO] Created Modulo APP stub at: "/Users/mmynsted/Library/Application Support/espanso/Modulo.app/Contents/MacOS/modulo"
16:37:47 [ INFO] worker is running!
16:37:47 [ INFO] watching for changes in path: /Users/mmynsted/Library/Preferences/espanso
16:39:24 [ INFO] Received signal: 2, terminating worker
16:39:24 [ INFO] terminating espanso.
```

**Comment by mmynsted** [comment]
If I do the following, espanso starts and stays running so long as I do not kill it.

```bash
$ espanso daemon
```

`espanso start` appears to start then stop.

**Comment by federico-terzi** [comment]
@mmynsted Perhaps this has to do with a root registration, could you please try with:

```
sudo espanso unregister
espanso register
```

And see if that helps? Perhaps the service was registered with root permissions and that is interfering

**Comment by mmynsted** [comment]
Yeah, I did that already.

**Comment by mmynsted** [comment]
Okay. I have it working now. Wow, registering with `root` really caused some issues.

The following fixed it.

```bash
sudo rm /tmp/espanso.out /tmp/espanso.err
```

**Comment by federico-terzi** [comment]
@mmynsted Great! Glad you got that working and thank you for the investigation. 

I'll make sure to add a check to prevent registering with root permissions on macOS :)

**Comment by federico-terzi** [comment]
Hey @tobiasbueschel,

I'm sorry you experienced this problem! I haven't fixed this bug yet because the new Espanso version is an app bundle now, which should highly mitigate this issue :)

> Then I tried to restart my machine to see if that would start the service - but to my unfortunate surprise, something went wrong with the service and now my MacBook crashes each time I before starting to boot into my user.

That's really strange! I'm sorry about that

> Hence, I wanted to ask what other files would need to be removed and how I can cleanly remove anything installed by Espanso?

So here's what I would do:
* In safe mode, open the `/Users/<your user here>/Library/LaunchAgents` directory and delete the `com.federicoterzi.espanso.plist` file
* Uninstall espanso with `brew remove espanso`
* Delete the `/Users/<your user here>/Application Support/espanso` directory
* Restart your mac

Let me know if that helps

Cheers :)

**Comment by federico-terzi** [comment]
Hey,

I'm publishing a fix in the upcoming v2.0.5-alpha release, thank you for the help!

Cheers :)
